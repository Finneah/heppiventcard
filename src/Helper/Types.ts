export type StampCardType = {
  id?: string;
  title: string;
  complete: Boolean;
  date_of_creation: Date;
  date_of_completed?: Date;
  completed_image?: string;
};

export interface SectionPart extends StampCardType {
  content: StampType[] | null | undefined;
}

export type StampType = {
  id?: string;
  number: number;
  name: string | null | undefined;
  image: string | null | undefined;
  picture: string | null | undefined;
  done: Boolean;
  date: Date | undefined;
  description: string | null | undefined;
  stampCard: StampCardType;
};

export type UserType = {
  id?: number;
  name: string | null | undefined;
  rank: number;
};

export type QRCodeObject = {
  url: string;
  date: string;
  description: string;
  name: string;
  picture: string | null | undefined;
};
export type QRCodeReturnData = {
  image: string;
  description: string;
  date: Date;
  done: boolean;
  name: string;
};
