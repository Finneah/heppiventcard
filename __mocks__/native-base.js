jest.mock('native-base', () => {
  return {
    StyleProvider: jest.fn().mockImplementation((x) => x),
  };
});
