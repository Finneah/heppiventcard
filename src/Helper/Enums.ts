export enum DEBUGLEVEL {
  OFF,
  DATA,
  ALL,
}
export enum UserRank {
  RANK_00 = 0,
  RANK_03 = 3,
  RANK_10 = 10,
  RANK_15 = 15,
  RANK_20 = 20,
  RANK_30 = 30,
}
