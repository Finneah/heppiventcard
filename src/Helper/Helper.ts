import {StampCards, Stamps} from '../database';
import {UserRank} from './Enums';
import {strings} from '../locale/i18n';
import {SectionPart, StampCardType, UserType} from './Types';
import moment from 'moment';
import 'moment/locale/de';
export const isValidUser = (user: UserType): Boolean => {
  try {
    if (user?.name !== '' || user?.rank >= 0) {
      return true;
    }
    throw new Error('name or rank is missing');
  } catch (error) {
    return false;
  }
};
/**
 * "RANK_0": "Gast",
 * "RANK_3": "Stammgast",
 * "RANK_10": "Mitglied",
 * "RANK_15": "Stammmitglied",
 * "RANK_20": "Platinmitglied",
 * "RANK_30": "Goldmitglied",
 */
export const _getRank = (user: UserType): string => {
  let returnValue = strings('RANK_00');
  if (user?.rank) {
    if (user.rank >= UserRank.RANK_30) {
      returnValue = strings('RANK_30');
    } else if (user.rank >= UserRank.RANK_20) {
      returnValue = strings('RANK_20');
    } else if (user.rank >= UserRank.RANK_15) {
      returnValue = strings('RANK_15');
    } else if (user.rank >= UserRank.RANK_10) {
      returnValue = strings('RANK_10');
    } else if (user.rank >= UserRank.RANK_03) {
      returnValue = strings('RANK_03');
    } else {
      returnValue = strings('RANK_00');
    }
  }
  return returnValue;
};

export const _getImageNameForRank = (user: UserType): string => {
  if (user?.rank >= 30) {
    return 'Gold';
  } else if (user.rank >= 20) {
    return 'Platin';
  }
  return 'Standard';
};

export const _checkDoneItems = (stampCard: SectionPart): number | string => {
  try {
    var done = 0;

    if (stampCard?.content) {
      stampCard?.content?.forEach((content: any) => {
        if (content.done === 1 || content.done === true) {
          done++;
        }
      });

      return done;
    } else {
      throw new Error('Error on _checkDoneItems');
    }
  } catch (error) {
    console.warn('_checkDoneItems', error);
    return '_checkDoneItems' + ' ' + error.message;
  }
};

export const _getStampCardTitle = (stampCard: SectionPart): string => {
  let firstDate: Date | undefined;
  let lastDate: Date | undefined;
  let title = '';

  if (stampCard?.content) {
    for (const element of stampCard.content) {
      if (!firstDate && element.date) {
        firstDate = element.date;
      } else {
        if (
          firstDate &&
          element?.date !== null &&
          element?.date !== undefined &&
          element?.date < firstDate
        ) {
          firstDate = element.date;
        }
      }
      if (!lastDate) {
        lastDate = element.date;
      } else {
        if (
          lastDate &&
          element?.date !== null &&
          element?.date !== undefined &&
          element?.date > lastDate
        ) {
          lastDate = element.date;
        }
      }

      if (firstDate && lastDate) {
        title =
          moment(firstDate).locale('de').format('L') +
          ' - ' +
          moment(lastDate).locale('de').format('L');
        console.log(title);
      }
    }
  } else {
    return title;
  }

  return title;
};

export const _createNewStampCard = async () => {
  var newStampCard: StampCardType = {
    title: '',
    complete: false,
    date_of_creation: new Date(),
    date_of_completed: undefined,
    completed_image: undefined,
  };

  newStampCard.date_of_creation = new Date();
  newStampCard.title = '';
  newStampCard.complete = false;

  var card = await StampCards.insert(newStampCard, true)[0];

  var newStamps = [];
  for (let i = 1; i <= 10; i++) {
    newStamps.push({
      number: i,
      done: 0,
      stampCard: card,
    });
  }

  await Stamps.insert(newStamps, true);
};
