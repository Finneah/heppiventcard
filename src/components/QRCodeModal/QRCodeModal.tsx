/* eslint-disable react-native/no-inline-styles */
import {Heading, Icon, Pressable, Spinner, Modal, useToast} from 'native-base';
import {StyleSheet, Dimensions, Vibration} from 'react-native';
import * as React from 'react';
import Ionicons from 'react-native-vector-icons/Ionicons';
import {QRCodeReturnData, SectionPart, StampType} from '../../Helper/Types';
import GlobalColors from '../../styles/GlobalColors';
import {strings} from '../../locale/i18n';
import QRCodeScanner from 'react-native-qrcode-scanner';

import {
  _createQRCodeDataObject,
  _getImageFromWebsite,
  _getStampExists,
} from './QRCodeModalHelper';
import {_checkDoneItems} from '../../Helper/Helper';

type Props = {
  stampCard: SectionPart;
  stamp: StampType;
  index: number;

  _onSuccessReadQRCodeCallback: Function;
};
const numCol = 3;

const QRCodeModal: React.FC<Props> = ({
  stampCard,
  stamp,
  index,
  _onSuccessReadQRCodeCallback,
}) => {
  const [qrCodeModalVisible, setQRCodeModalVisible] = React.useState(false);
  const [showLoadingSpinner, setShowLoadingSpinner] = React.useState(false);

  const toast = useToast();

  React.useEffect(() => {
    if (!qrCodeModalVisible) {
      setQRCodeModalVisible(false);
      setShowLoadingSpinner(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [qrCodeModalVisible]);

  function setAndShowToast(
    title: string,
    description: string,
    status: 'error' | 'info' | 'warning' | 'success' | undefined,
  ) {
    setShowLoadingSpinner(false);
    setQRCodeModalVisible(false);

    toast.show({
      title: title,
      description: description,
      placement: 'bottom',
      status: status,
    });
  }

  async function _onSuccessReadQRCode(e: any) {
    try {
      setShowLoadingSpinner(true);
      Vibration.vibrate([500, 500], false);

      var data = _createQRCodeDataObject(e.data);

      var image;

      try {
        image = await _getImageFromWebsite(data);
        // @todo get Picture
      } catch (error) {
        console.info(error);
        setShowLoadingSpinner(false);
        setAndShowToast(
          strings('SRY'),
          strings('CANT_READ_QR_CODE' + ' ' + error),
          'error',
        );
      }

      if (image) {
        var exist = _getStampExists(data);
        if (__DEV__) {
          exist = false;
        }

        if (exist) {
          setAndShowToast(
            strings('DO_NOT_CHEAT'),
            strings('STAMP_EXIST'),
            'info',
          );
        } else {
          let qrCodeData: QRCodeReturnData = {
            image: image,
            // picture:
            description: data.description,
            date: new Date(data.date),
            done: true,
            name: data.name,
          };

          setShowLoadingSpinner(false);
          setQRCodeModalVisible(false);

          _onSuccessReadQRCodeCallback(stamp, qrCodeData);
        }
      } else {
        setShowLoadingSpinner(false);
        setAndShowToast(strings('SRY'), strings('CANT_READ_QR_CODE'), 'error');
      }
    } catch (error) {
      setShowLoadingSpinner(false);
      setAndShowToast(strings('SRY'), strings('CANT_READ_QR_CODE'), 'error');
    }
  }

  return (
    <>
      <Pressable
        onPress={() => {
          _checkDoneItems(stampCard) === index
            ? setQRCodeModalVisible(true)
            : null;
        }}
        key={(stamp.number + index).toString()}
        style={[styles.item, styles.stampItem]}>
        {_checkDoneItems(stampCard) === index ? (
          <Icon color="primary.600" as={<Ionicons name="add" />} />
        ) : (
          <Heading size="md" style={{color: GlobalColors.brandPrimary}}>
            {stamp.number.toString()}
          </Heading>
        )}
      </Pressable>
      <Modal
        size="full"
        isOpen={qrCodeModalVisible}
        onClose={() => setQRCodeModalVisible(false)}>
        <Modal.Content style={styles.bottomModal} w={'100%'}>
          <Modal.CloseButton />
          <Modal.Header>
            <Heading style={{color: GlobalColors.dark}}>
              {strings('SCAN_QR_CODE_NOW')}
            </Heading>
          </Modal.Header>
          <Modal.Body>
            <QRCodeScanner
              reactivate={true}
              reactivateTimeout={5000}
              showMarker={true}
              markerStyle={{
                borderColor: GlobalColors.brandSecondary,
                borderRadius: 25,
              }}
              cameraStyle={{
                alignSelf: 'center',
                width: '100%',
              }}
              customMarker={
                showLoadingSpinner ? (
                  <Spinner
                    accessibilityLabel="scanning QRCode"
                    color="secondary.400"
                  />
                ) : undefined
              }
              fadeIn={true}
              onRead={_onSuccessReadQRCode.bind(this)}
              topContent={undefined}
              bottomContent={undefined}
            />
          </Modal.Body>
        </Modal.Content>
      </Modal>
    </>
  );
};
const styles = StyleSheet.create({
  item: {
    zIndex: 0,
    flexDirection: 'column',
    backgroundColor: GlobalColors.lightGrey,
    margin: 5,
    padding: 0,
    borderRadius: 20,
    width: Dimensions.get('screen').width * ((100 / numCol - 10) / 100),
    height: Dimensions.get('screen').width * ((100 / numCol - 10) / 100),
  },
  bottomModal: {
    marginBottom: 0,
    marginTop: 'auto',
  },
  stampItem: {
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
export default QRCodeModal;
