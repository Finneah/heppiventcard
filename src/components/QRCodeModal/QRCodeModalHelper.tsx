import RNFetchBlob from 'rn-fetch-blob';
import {StampsModel} from '../../database/Models/StampsModel';

import {QRCodeObject, StampType} from '../../Helper/Types';

let stampsModel = new StampsModel();

export const _createQRCodeDataObject = (qrCodeData: string): QRCodeObject => {
  var data: QRCodeObject = {
    url: '',
    date: '',
    description: '',
    name: '',
    picture: undefined,
  };

  qrCodeData.split('&').forEach((element: any) => {
    if (element) {
      if (element.indexOf('url=') !== -1) {
        data.url = element.replace('url=', '').replace('\n', '');
      } else if (element.indexOf('date=') !== -1) {
        data.date = element.replace('date=', '').replace('\n', '');
      } else if (element.indexOf('description=') !== -1) {
        data.description = element
          .replace('description=', '')
          .replace('\n', '');
      } else if (element.indexOf('name=') !== -1) {
        data.name = element.replace('name=', '').replace('\n', '');
      } else if (element.indexOf('picture=') !== -1) {
        data.picture = element.replace('picture=', '').replace('\n', '');
      }
    }
  });

  return data;
};

export const _getImageFromWebsite = async (
  data: any,
): Promise<string | void> => {
  try {
    var res = await new Promise<any>((resolve) =>
      resolve(RNFetchBlob.fetch('GET', data.url, {})),
    );

    let status = res?.info()?.status;
    if (status === 200) {
      return res?.data;
    } else {
      throw new Error('Image not found');
    }
  } catch (error) {
    throw new Error('Image not found' + ' ' + error.message);
  }
};

export const _getStampExists = (data: any): boolean => {
  var exist: StampType[] = stampsModel.filterStampsBy({
    description: data.description,
    date: new Date(data.date),
    name: data.name,
  });

  if (exist.length !== 0) {
    return true;
  }
  return false;
};
