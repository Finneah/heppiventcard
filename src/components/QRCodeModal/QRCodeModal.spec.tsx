import 'react-native';
import * as React from 'react';
import {render} from '@testing-library/react-native';

// Note: test renderer must be required after react-native.

import {NativeBaseProvider} from 'native-base';

import QRCodeModal from './QRCodeModal';
import {SectionPart, StampCardType, StampType} from '../../Helper/Types';

const wrapper = ({children}: any) => (
  <NativeBaseProvider
    initialWindowMetrics={{
      frame: {x: 0, y: 0, width: 0, height: 0},
      insets: {top: 0, left: 0, right: 0, bottom: 0},
    }}>
    {children}
  </NativeBaseProvider>
);

describe('HomeScreen', () => {
  test('renders correctly', () => {
    let stampCard: StampCardType = {
      title: 'string',
      complete: false,
      date_of_creation: new Date(),
      date_of_completed: undefined,
      completed_image: undefined,
    };

    let stamp: StampType = {
      number: 1,
      name: undefined,
      image: undefined,
      done: false,
      picture: undefined,
      date: new Date(),
      description: 'string',
      stampCard: stampCard,
    };

    let content: StampType[] = [stamp];
    let item: SectionPart = {
      id: 'string',
      title: 'string',
      complete: false,
      date_of_creation: new Date(),
      date_of_completed: undefined,
      completed_image: 'string',
      content: content,
    };
    const qrCodeModal = render(
      <QRCodeModal
        stampCard={item}
        index={0}
        stamp={stamp}
        _onSuccessReadQRCodeCallback={(result: any) => console.log(result)}
      />,
      {wrapper},
    );

    expect(qrCodeModal.toJSON()).toMatchSnapshot();
  });
});
