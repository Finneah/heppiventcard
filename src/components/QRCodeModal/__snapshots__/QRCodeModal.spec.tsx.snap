// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onHoverIn={[Function]}
    onHoverOut={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {},
        Object {
          "backgroundColor": "#F8F8F8",
          "borderRadius": 20,
          "flexDirection": "column",
          "height": 175.00000000000003,
          "margin": 5,
          "padding": 0,
          "width": 175.00000000000003,
          "zIndex": 0,
        },
        Object {
          "alignItems": "center",
          "borderRadius": 20,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      fontSize="32px"
      lineHeight="32px"
      style={
        Array [
          Object {
            "color": "primary.600",
            "fontSize": 12,
          },
          Array [
            Object {
              "color": "#0891b2",
              "fontSize": 32,
              "height": 32,
              "lineHeight": 32,
              "width": 32,
            },
          ],
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ï„€
    </Text>
  </View>
  <View
    alignItems="center"
    display="flex"
    justifyContent="center"
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
</RNCSafeAreaProvider>
`;
